samples/web/content/apprtc/js/call.js

```diff
@@ -0,0 +1,364 @@
+    return;
+  }
+
+  if (this.pcClient_) {
+    this.pcClient_.close();
+    this.pcClient_ = null;
+  }
+
+  // Send bye to GAE. This must complete before saying BYE to other client.
+  // When the other client sees BYE it attempts to post offer and candidates to
+  // GAE. GAE needs to know that we're disconnected at that point otherwise
+  // it will forward messages to this client instead of storing them.
+  var path = '/bye/' + this.params_.roomId + '/' + this.params_.clientId;
+  var xhr = new XMLHttpRequest();
+  xhr.open('POST', path, false);
+  xhr.send();
+
+  // Send bye to other client.
+  this.channel_.send(JSON.stringify({ type: 'bye' }));
+  this.channel_.close();
+
+  this.params_.roomId = null;
+  this.params_.clientId = null;
+};
+
+Call.prototype.onRemoteHangup = function() {
+  this.startTime = null;
+
+  // On remote hangup this client becomes the new initiator.
+  this.params_.isInitiator = true;
+
+  if (this.pcClient_) {
+    this.pcClient_.close();
+    this.pcClient_ = null;
+  }
+
+  this.startSignaling_();
+};
+
+Call.prototype.getPeerConnectionStates = function() {
+  if (!this.pcClient_) {
+    return null;
+  }
+  return this.pcClient_.getPeerConnectionStates();
+};
+
+Call.prototype.getPeerConnectionStats = function(callback) {
+  if (!this.pcClient_) {
+    return;
+  }
+  this.pcClient_.getPeerConnectionStats(callback);
+};
+
+Call.prototype.toggleVideoMute = function() {
+  var videoTracks = this.localStream_.getVideoTracks();
+  if (videoTracks.length === 0) {
+    trace('No local video available.');
+    return;
+  }
+
+  trace('Toggling video mute state.');
+  for (var i = 0; i < videoTracks.length; ++i) {
+    videoTracks[i].enabled = !videoTracks[i].enabled;
+  }
+
+  trace('Video ' + (videoTracks[0].enabled ? 'unmuted.' : 'muted.'));
+};
+
+Call.prototype.toggleAudioMute = function() {
+  var audioTracks = this.localStream_.getAudioTracks();
+  if (audioTracks.length === 0) {
+    trace('No local audio available.');
+    return;
+  }
+
+  trace('Toggling audio mute state.');
+  for (var i = 0; i < audioTracks.length; ++i) {
+    audioTracks[i].enabled = !audioTracks[i].enabled;
+  }
+
+  trace('Audio ' + (audioTracks[0].enabled ? 'unmuted.' : 'muted.'));
+};
+
+// Connects client to the room. This happens by simultaneously requesting
+// media, requesting turn, and registering with GAE. Once all three of those
+// tasks is complete, the signaling process begins. At the same time, a
+// WebSocket connection is opened using |wss_url| followed by a subsequent
+// registration once GAE registration completes.
+Call.prototype.connectToRoom_ = function(mediaPromise, turnPromise) {
+  // Asynchronously open a WebSocket connection to WSS.
+  // TODO(jiayl): We don't need to wait for the signaling channel to open before
+  // start signaling.
+  var channelPromise = this.channel_.open().catch(function(error) {
+    this.onError_('WebSocket open error: ' + error.message);
+    return Promise.reject(error);
+  });
```

Comment:
Does this function (the function passed to catch) need .bind(this) like the other ones in this function (connectToRoom)?